Automattic/jetpack
# Fix YouTube AMP compatibility and improve non-AMP YouTube embeds

Fixes #16946.

Given this `post_content`:

```html

<p>YouTube embed block</p>



<figure class="wp-block-embed-youtube wp-block-embed is-type-video is-provider-youtube wp-embed-aspect-16-9 wp-has-aspect-ratio"><div class="wp-block-embed__wrapper">
https://www.youtube.com/watch?v=BTbjxU3d918
</div></figure>



<p>Shortcode: videoseries</p>



[youtube https://www.youtube.com/videoseries?list=PL56C3506BBE979C1B]



<p>Shortcode: playlist</p>



[youtube https://www.youtube.com/playlist?list=PLAVZ4NFtZX0fE54mDSqNKym-o_rz-8xmk]



<p>Shortcode: video in playlist</p>



[youtube https://www.youtube.com/watch?v=GJNxoe-iSb4&list=PLAVZ4NFtZX0fE54mDSqNKym-o_rz-8xmk]

```

Take special note of the last embed. It is a playlist but it has one specific video (the 7th of the playlist) being specified. Therefore the iframe should start playing with the 7th video, not the first.

##### Before 👎 

Non-AMP with JS | Non-AMP sans JS | AMP with JS | AMP sans JS
------------------|--------------------|-------------|--------------
![image](https://user-images.githubusercontent.com/134745/94498911-2f0a2d00-01b0-11eb-8e06-d61760f212ce.png) | ![image](https://user-images.githubusercontent.com/134745/94498936-3df0df80-01b0-11eb-9c5e-a72670f0468b.png) | ![image](https://user-images.githubusercontent.com/134745/94498960-4ea15580-01b0-11eb-80c9-7a8dfa4e08bb.png) | ![image](https://user-images.githubusercontent.com/134745/94499003-6ed11480-01b0-11eb-9fd7-9bf9464fd111.png)

In the “AMP with JS” example, there is a JS error initially console:

> Uncaught DOMException: Failed to construct 'PresentationRequest': The document is sandboxed and lacks the 'allow-presentation' flag.

And when trying to click on the title of the first YouTube video:

> Blocked opening 'https://www.youtube.com/watch?v=BTbjxU3d918&feature=emb_title' in a new window because the request was made in a sandboxed frame whose 'allow-popups' permission is not set.

#### After 👍 

Non-AMP with JS | Non-AMP sans JS | AMP with JS | AMP sans JS
------------------|--------------------|-------------|--------------
![image](https://user-images.githubusercontent.com/134745/94499211-f7e84b80-01b0-11eb-82f1-62197b6b9848.png) | ![image](https://user-images.githubusercontent.com/134745/94499230-02a2e080-01b1-11eb-9f5e-53ab3ab3b939.png) | ![image](https://user-images.githubusercontent.com/134745/94499274-16e6dd80-01b1-11eb-9aae-9d57d136ef25.png) | ![image](https://user-images.githubusercontent.com/134745/94499295-236b3600-01b1-11eb-8ca9-019345c396b1.png)

Notice how in AMP and non-AMP alike, the last embed now currently shows the 7th video being queued up rather than the 1st.

When interacting with the YouTube embeds in AMP, no `sandbox` errors are now raised.

In the AMP sans JS screenshot, notice that the `noscript` fallback image link is being rendered instead of the broken iframe. Please also note that in this screenshot, SSR has been turned off: currently when SSR is enabled, nothing gets rendered for any of the embeds. This will be fixed by https://github.com/ampproject/amphtml/pull/29846. 

Also, when embedding YouTube playlists that do not have an individual video indicated, an `iframe` is used which will be converted into an `amp-iframe`. See also https://github.com/ampproject/amp-wp/issues/465.

#### Changes proposed in this Pull Request:

* Deprecate `jetpack_amp_youtube_shortcode()` in favor of making `youtube_id()` fully AMP-compatible.
* Embed YouTube videos using `amp-youtube` rather than `iframe` when an individual video is being referenced.
* Fix embedding YouTube videos that have associated playlists.

#### Jetpack product discussion



#### Does this pull request change what data or activity we track or use?



#### Testing instructions:






1. Activate AMP plugin.
2. Create post with YouTube embeds, like the `post_content` above.
3. Compare AMP and non-AMP.
4. Ensure no errors are in the console for YouTube.

#### Proposed changelog entry for your changes:

* Improve embedding of Jetpack videos, including on AMP pages and generally for videos with playlists.
